File: day06.c
#include<stdio.h>

int main(void){
    char T[160][160];
    FILE *f=fopen("data/input06.txt","r");
    int N=0;
    while(fgets(T[N],160,f)){N++;}
    int startx=0,starty=0;
    for(int y=0; y<N; y++){
        for(int x=0; x<N; x++){
            if(T[y][x]=='^'){startx=x; starty=y;}
        }
    }
    int walk(int ispart1){
        int x=startx,y=starty,dx=0,dy=-1;
        int steps=0;
        while(1){
            if(ispart1){T[y][x]='*';}
            int nx=x+dx,ny=y+dy;
            if(nx<0||nx>=N||ny<0||ny>=N){return 0;}
            if(steps>7000){return 1;}
            if(T[ny][nx]=='#'){int temp=dx;dx=-dy;dy=temp;}else{x=nx;y=ny;steps++;}
        }
    }
    int part1=0,part2=0;
    walk(1);
    for(int y=0; y<N; y++){
        for(int x=0; x<N; x++){
            if(T[y][x]=='*' && (x!=startx || y!=starty)){
                part1++; 
                T[y][x]='#'; part2+=walk(0); T[y][x]='.';
            }
        }
    }
    printf("%d %d\n",part1,part2);
    return 0;
}

File: day06.lua
local part1,part2=0,0
W,H,startx,starty=0,0,0,0
grid={}
for line in io.lines("data/input06.txt") do
    H=H+1
    table.insert(grid,{})
    for i=1,#line do
        local ch = string.sub(line,i,i)
        table.insert(grid[H],ch)
        if ch=="^" then startx,starty=i,H end
    end
end
W=H
function walk()
    local x,y=startx,starty
    local h,dx,dy=0,0,-1
    seenpos={[200*y+x]=1}
    seenstates={[4*(200*y+x)+h]=1}
    while true do
        local nx,ny=x+dx,y+dy
        if nx<1 or nx>W or ny<1 or ny>H then return false end
        if grid[ny][nx]=="#" then
            dx,dy=-dy,dx
            h=(h+1)%4
        else
            x,y=nx,ny
            seenpos[200*y+x]=1
            if seenstates[4*(200*y+x)+h] then
                return true
            else
                seenstates[4*(200*y+x)+h]=1
            end
        end
    end
end

walk()
for _,k in pairs(seenpos) do
    part1=part1+1
    grid[_//200][_%200]="#"
    if walk(false) then part2=part2+1 end
    grid[_//200][_%200]="."
end
print(part1,part2)